File: shared/runner/runner.py
"""
Rate-limit-aware real server runner.

Respects:
  - 60s cooldown between games
  - 40 games/hour max
  - 300 games/day max

Usage:
  python runner.py \
    --tokens easy:TOKEN1 medium:TOKEN2 hard:TOKEN3 \
    --difficulties easy medium hard \
    --games-per-day 200 \
    --bot-dir ../../challenges/groceries/bot

Tokens are single-use — you need a fresh token from app.ainm.no/challenge for each game.
This script just enforces rate limits and calls client.py for each game.
"""
from __future__ import annotations
import subprocess
import time
import logging
import argparse
from datetime import datetime, timedelta
from collections import deque
from pathlib import Path

logger = logging.getLogger("runner")
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
)

COOLDOWN_SECONDS = 61       # 60s minimum + 1s buffer
MAX_GAMES_PER_HOUR = 38     # 40 - 2 buffer
MAX_GAMES_PER_DAY = 280     # 300 - 20 buffer for manual games


class RateLimitedRunner:
    def __init__(
        self,
        bot_script: str,
        games_per_day: int = MAX_GAMES_PER_DAY,
    ):
        self.bot_script = bot_script
        self.games_per_day = min(games_per_day, MAX_GAMES_PER_DAY)
        self.game_times: deque[float] = deque()  # timestamps of recent games

    def _wait_if_needed(self):
        now = time.time()

        # Enforce cooldown since last game
        if self.game_times:
            elapsed = now - self.game_times[-1]
            if elapsed < COOLDOWN_SECONDS:
                wait = COOLDOWN_SECONDS - elapsed
                logger.info(f"Cooldown: waiting {wait:.1f}s...")
                time.sleep(wait)
                now = time.time()

        # Enforce hourly limit
        one_hour_ago = now - 3600
        recent_hour = sum(1 for t in self.game_times if t > one_hour_ago)
        if recent_hour >= MAX_GAMES_PER_HOUR:
            oldest_in_window = min(t for t in self.game_times if t > one_hour_ago)
            wait = (oldest_in_window + 3600) - now + 5
            logger.info(f"Hourly limit reached. Waiting {wait:.0f}s ({wait/60:.1f}min)...")
            time.sleep(max(wait, 0))
            now = time.time()

        # Enforce daily limit
        one_day_ago = now - 86400
        recent_day = sum(1 for t in self.game_times if t > one_day_ago)
        if recent_day >= self.games_per_day:
            logger.warning(f"Daily limit reached ({recent_day}/{self.games_per_day}). Stopping.")
            raise SystemExit(0)

    def run_game(self, token: str, difficulty: str) -> int:
        """Runs a single game and returns the final score."""
        self._wait_if_needed()

        logger.info(f"Starting game: {difficulty}")
        t0 = time.time()

        result = subprocess.run(
            ["python", self.bot_script, "--token", token, "--difficulty", difficulty],
            capture_output=True,
            text=True,
            timeout=400,  # max game time with buffer
        )

        elapsed = time.time() - t0
        self.game_times.append(time.time())

        if result.returncode != 0:
            logger.error(f"Game failed: {result.stderr[-500:]}")
            return 0

        # Extract score from last line of output
        score = 0
        for line in reversed(result.stdout.strip().split("\n")):
            if "Final score:" in line:
                try:
                    score = int(line.split("Final score:")[-1].strip())
                except ValueError:
                    pass
                break

        logger.info(f"Game complete: {difficulty} | Score: {score} | Time: {elapsed:.0f}s")
        return score

    def run_schedule(self, schedule: list[tuple[str, str]]):
        """
        Runs a list of (token, difficulty) tuples with rate limiting.
        schedule = [("TOKEN_A", "easy"), ("TOKEN_B", "medium"), ...]
        """
        results = []
        for i, (token, difficulty) in enumerate(schedule):
            logger.info(f"Game {i+1}/{len(schedule)}: {difficulty}")
            score = self.run_game(token, difficulty)
            results.append({"difficulty": difficulty, "score": score, "timestamp": datetime.now().isoformat()})

        # Summary
        logger.info("\n=== Run Summary ===")
        for r in results:
            logger.info(f"  {r['difficulty']:12s}: {r['score']}")
        return results


def main():
    parser = argparse.ArgumentParser(description="NM i AI 2026 — Rate-Limited Runner")
    parser.add_argument("--bot-dir", required=True, help="Path to bot directory")
    parser.add_argument(
        "--games",
        nargs="+",
        metavar="DIFFICULTY:TOKEN",
        help="Games to run as difficulty:token pairs, e.g. easy:JWT_TOKEN_HERE",
    )
    parser.add_argument("--games-per-day", type=int, default=MAX_GAMES_PER_DAY)
    args = parser.parse_args()

    bot_script = str(Path(args.bot_dir) / "client.py")
    runner = RateLimitedRunner(bot_script=bot_script, games_per_day=args.games_per_day)

    schedule = []
    for game_spec in (args.games or []):
        parts = game_spec.split(":", 1)
        if len(parts) == 2:
            schedule.append((parts[1], parts[0]))  # (token, difficulty)
        else:
            logger.warning(f"Invalid game spec (expected difficulty:token): {game_spec}")

    if not schedule:
        logger.error("No games specified. Use --games easy:TOKEN medium:TOKEN ...")
        return

    runner.run_schedule(schedule)


if __name__ == "__main__":
    main()

File: shared/runner/runner.ts
import { execFileSync } from "node:child_process";
import minimist from "minimist";

const COOLDOWN_SECONDS = 61;
const MAX_GAMES_PER_HOUR = 38;
const MAX_GAMES_PER_DAY = 280;

function log(level: string, message: string): void {
  const timestamp = new Date().toISOString();
  console.error(`${timestamp} [${level}] ${message}`);
}

function sleep(ms: number): void {
  Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, ms);
}

class RateLimitedRunner {
  private botScript: string;
  private gamesPerDay: number;
  private gameTimes: number[] = [];

  constructor(botScript: string, gamesPerDay = MAX_GAMES_PER_DAY) {
    this.botScript = botScript;
    this.gamesPerDay = Math.min(gamesPerDay, MAX_GAMES_PER_DAY);
  }

  private waitIfNeeded(): void {
    let now = Date.now() / 1000;

    // Enforce cooldown since last game
    if (this.gameTimes.length > 0) {
      const elapsed = now - this.gameTimes[this.gameTimes.length - 1];
      if (elapsed < COOLDOWN_SECONDS) {
        const wait = COOLDOWN_SECONDS - elapsed;
        log("INFO", `Cooldown: waiting ${wait.toFixed(1)}s...`);
        sleep(Math.ceil(wait * 1000));
        now = Date.now() / 1000;
      }
    }

    // Enforce hourly limit
    const oneHourAgo = now - 3600;
    const recentHour = this.gameTimes.filter((t) => t > oneHourAgo).length;
    if (recentHour >= MAX_GAMES_PER_HOUR) {
      const oldestInWindow = Math.min(
        ...this.gameTimes.filter((t) => t > oneHourAgo),
      );
      const wait = oldestInWindow + 3600 - now + 5;
      log("INFO", `Hourly limit reached. Waiting ${wait.toFixed(0)}s (${(wait / 60).toFixed(1)}min)...`);
      sleep(Math.max(Math.ceil(wait * 1000), 0));
      now = Date.now() / 1000;
    }

    // Enforce daily limit
    const oneDayAgo = now - 86400;
    const recentDay = this.gameTimes.filter((t) => t > oneDayAgo).length;
    if (recentDay >= this.gamesPerDay) {
      log("WARN", `Daily limit reached (${recentDay}/${this.gamesPerDay}). Stopping.`);
      process.exit(0);
    }
  }

  runGame(token: string, difficulty: string): number {
    this.waitIfNeeded();

    log("INFO", `Starting game: ${difficulty}`);
    const t0 = Date.now();

    try {
      const result = execFileSync(
        "node",
        ["--experimental-transform-types", this.botScript, "--token", token, "--difficulty", difficulty],
        { timeout: 400_000, encoding: "utf-8" },
      );

      const elapsed = (Date.now() - t0) / 1000;
      this.gameTimes.push(Date.now() / 1000);

      // Extract score from last line of output
      let score = 0;
      const lines = result.trim().split("\n");
      for (let idx = lines.length - 1; idx >= 0; idx--) {
        if (lines[idx].includes("Final score:")) {
          const match = lines[idx].split("Final score:")[1]?.trim();
          if (match) score = parseInt(match, 10) || 0;
          break;
        }
      }

      log("INFO", `Game complete: ${difficulty} | Score: ${score} | Time: ${elapsed.toFixed(0)}s`);
      return score;
    } catch (err: any) {
      const elapsed = (Date.now() - t0) / 1000;
      this.gameTimes.push(Date.now() / 1000);
      log("ERROR", `Game failed: ${err.stderr?.slice(-500) ?? err.message}`);
      return 0;
    }
  }

  runSchedule(schedule: [string, string][]): Record<string, any>[] {
    const results: Record<string, any>[] = [];
    for (let idx = 0; idx < schedule.length; idx++) {
      const [token, difficulty] = schedule[idx];
      log("INFO", `Game ${idx + 1}/${schedule.length}: ${difficulty}`);
      const score = this.runGame(token, difficulty);
      results.push({
        difficulty,
        score,
        timestamp: new Date().toISOString(),
      });
    }

    log("INFO", "\n=== Run Summary ===");
    for (const result of results) {
      log("INFO", `  ${result.difficulty.padEnd(12)}: ${result.score}`);
    }
    return results;
  }
}

function main(): void {
  const args = minimist(process.argv.slice(2), {
    string: ["bot-dir", "games"],
    default: { "games-per-day": MAX_GAMES_PER_DAY },
  });

  if (!args["bot-dir"]) {
    console.error("Usage: node --experimental-transform-types runner.ts --bot-dir DIR --games difficulty:TOKEN ...");
    process.exit(1);
  }

  const botScript = `${args["bot-dir"]}/client.ts`;
  const runner = new RateLimitedRunner(botScript, args["games-per-day"]);

  const gameSpecs: string[] = Array.isArray(args.games) ? args.games : args.games ? [args.games] : [];
  // Also grab any remaining positional args after --games
  const allSpecs = [...gameSpecs, ...(args._ as string[])];

  const schedule: [string, string][] = [];
  for (const spec of allSpecs) {
    const parts = spec.split(":", 2);
    if (parts.length === 2) {
      schedule.push([parts[1], parts[0]]); // [token, difficulty]
    } else {
      log("WARN", `Invalid game spec (expected difficulty:token): ${spec}`);
    }
  }

  if (schedule.length === 0) {
    log("ERROR", "No games specified. Use --games easy:TOKEN medium:TOKEN ...");
    process.exit(1);
  }

  runner.runSchedule(schedule);
}

main();
